R/source.R

```diff
@@ -282,22 +269,117 @@ invoke_cargo <- function(toolchain, specific_target, dir, profile,
-      sprintf("--manifest-path %s", file.path(dir, "Cargo.toml")),
-      sprintf("--target-dir %s", file.path(dir, "target")),
-      if (profile == "release") "--release" else NULL
+      glue("--target={specific_target}"),
+      glue("--manifest-path={file.path(dir, 'Cargo.toml')}"),
+      glue("--target-dir={file.path(dir, 'target')}"),
+      if (profile == "release") "--release" else NULL,
+      "--message-format=json-diagnostic-rendered-ansi",
+      if (tty_has_colors) {
+        "--color=always"
+      } else {
+        "--color=never"
+      }
     ),
-    stdout = stdout,
-    stderr = stderr
+    echo_cmd = FALSE,
+    windows_verbatim_args = FALSE,
+    stderr = ifelse(isTRUE(quiet), "|", ""),
+    stdout = "|",
+    error_on_status = FALSE,
+    stdout_line_callback = function(line, ...) {
+      assign("message_buffer", c(message_buffer, line), envir = env)
+    }
+  )
+
+  check_cargo_output(compilation_result, message_buffer, tty_has_colors, quiet)
+}
+
+#' Gathers ANSI-formatted cargo output
+#'
+#' Checks the output of cargo and filters messages according to `level`.
+#' Retrieves rendered ANSI strings and prepares them
+#' for `cli` and `glue` formatting.
+#' @param json_output \[ JSON(n) \] JSON messages produced by cargo.
+#' @param level \[ string \] Log level.
+#' Commonly used values are `"error"` and `"warning"`.
+#' @param tty_has_colors \[ logical(1) \] Indicates if output
+#' supports ANSI sequences. If `FALSE`, ANSI sequences are stripped off.
+#' @return \[ character(n) \] Vector of strings
+#' that can be passed to `ui_*`, `cli` or `glue` functions.
+#' @noRd
+gather_cargo_output <- function(json_output, level, tty_has_colors) {
+  rendered_output <-
+    json_output %>%
+    purrr::keep(
+      ~ .x$reason == "compiler-message" && .x$message$level == level
+    ) %>%
+    purrr::map_chr(~ .x$message$rendered)
+
+  if (!tty_has_colors) {
+    rendered_output <- cli::ansi_strip(rendered_output)
+  }
+
+  stringi::stri_replace_all_fixed(
+    rendered_output,
+    pattern = c("{", "}"),
+    replacement = c("{{", "}}"),
+    vectorize_all = FALSE
   )
-  if (status != 0L) {
-    ui_throw("Rust code could not be compiled successfully. Aborting.")
+}
+
+#' Processes output of cargo compilation process.
+#'
+#' Displays warnings emitted by the compiler
+#' and throws errors if compilation was unsuccessful.
+#' @param compilation_result The output of `processx::run()`.
+#' @param message_buffer \[ character(n) \] Messages emitted by cargo to stdout.
+#' @param tty_has_colors \[ logical(1) \] Indicates if output
+#' supports ANSI sequences. If `FALSE`, ANSI sequences are stripped off.
+#' @param quiet Logical indicating whether compile output should be generated or not.
+#' @noRd
+check_cargo_output <- function(compilation_result, message_buffer, tty_has_colors, quiet) {
+  cargo_output <- message_buffer %>%
+    purrr::map_chr(stringi::stri_trim) %>%
```

Comment:
Maybe we can parse JSONs without trimming whitespaces? Just curious.

Reply:
You were right, it seems to work fine.